[weibinlim] iP

weibinlim opened the pull request:
### Bigdog

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Bigdog frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/weibinlim/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
5. And it is FREE!

Features:

- [x]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

weibinlim committed:
Level 0. Rename, Greet, Exit

weibinlim committed:
Level 1: Echo

weibinlim committed:
Level 2: Add, List

weibinlim committed:
Level 3: Mark as Done

weibinlim committed:
Level 4: Todos, Events, Deadlines

weibinlim committed:
A-TextUiTesting: Automated Text UI Testing

weibinlim committed:
Level 5: Handle Errors

weibinlim committed:
Level 6: Delete

weibinlim committed:
feat: Add FileSaver.java

Class that handles the writing and reading from the external file on the hard drive.
Features:
- IO read and format string from external file
- IO write and format string to external file

feat: Add Bigdog.txt

Bigdog.txt is the file in which the list will be saved to in the hard drive.

Add variations of factory method of()

In the classes Task, Deadline, Todo, Event add new factory methods to take in a boolean to indicate if task is marked or not, on top of the task description.

Task.java: Change constructor to take in additional boolean parameter

Todo.java: Change constructor to take in additional boolean parameter

Deadline.java: Change constructor to take in additional boolean parameter

Event.java: Change constructor to take in additional boolean parameter

Deadline.java: Reformat variables end and toString() method

Event.java: Reformat variables start and end and toString() method

Bigdog.java: Change toDoList to private

Bigdog.java: Remove comment

weibinlim committed:
Revert "Level 6: Delete"

This reverts commit c22070f84fc6edafcd008a3a4b89859f55acdb0d.

weibinlim committed:
feat: Add FileSaver.java

Class that handles the writing and reading from the external file on the hard drive.
Features:
- IO read and format string from external file
- IO write and format string to external file
- Error handling for IOException and FIleNotFoundException

feat: Add Bigdog.txt

Bigdog.txt is the file in which the list will be saved to in the hard drive.

weibinlim committed:
Add variations of factory method of()

In the classes Task, Deadline, Todo, Event add new factory methods to take in a boolean to indicate if task is marked or not, on top of the task description.

Task.java: Change constructor to take in additional boolean parameter

Todo.java: Change constructor to take in additional boolean parameter

Deadline.java: Change constructor to take in additional boolean parameter

Event.java: Change constructor to take in additional boolean parameter

Deadline.java: Reformat variables end and toString() method

Event.java: Reformat variables start and end and toString() method

weibinlim committed:
Bigdog.java: Change toDoList to private

Bigdog.java: Remove comment

Bigdog.java: Used new method in FileSaver class

weibinlim committed:
input.txt: Changed test input
EXPECTED.txt: Changed expected output

weibinlim committed:
Update error message

weibinlim committed:
Deadline.java: Change end variable from type String to type LocalDateTime

Deadline.java: Add new method to convert String input to LocalDateTime object

Deadline.java: Edit toString() method to return appropriate representation

Deadline.java: Update exception message

Deadline.java: Edit constructor to take in LocalDateTime object instead of String

Event.java: Change start and end variables from type String to type LocalDateTime

Event.java: Add new method stringToDate(...) to convert String input to LocalDateTime object

Event.java: Edit toString() method to return appropriate representation

Event.java: Update exception message

Event.java: Edit constructor to take in LocalDateTime object instead of String

Nicholaskoh1 commented on src/main/java/Bigdog.java:
```diff
@@ -0,0 +1,83 @@
+import java.util.Scanner;
```
Perhaps you could package the java files?

Nicholaskoh1 commented on src/main/java/Bigdog.java:
```diff
@@ -0,0 +1,83 @@
+    private static FileSaver fileSaver = new FileSaver("./src/main/Bigdog.txt");
+
+    private static void editList(String str) throws BigdogException {
+        if (str.isEmpty()) {
+            throw new BigdogException("Please give me something to add!");
+        }
+        if (str.startsWith("mark")) {
+            if (Integer.parseInt(str.substring(5)) <= 0 || Integer.parseInt(str.substring(5)) > toDoList.size()) {
+                throw new BigdogException("That's out of your list!");
+            }
+            int ind = Integer.parseInt(str.substring(5)) - 1;
+            toDoList.get(ind).mark();
+            System.out.print("Nice! I've marked this task as done:\n" + toDoList.get(ind) + "\n");
+        } else if (str.startsWith("unmark")) {
+            if (Integer.parseInt(str.substring(7)) <= 0 || Integer.parseInt(str.substring(7)) > toDoList.size()) {
+                throw new BigdogException("That's out of your list!");
+            }
+            int ind = Integer.parseInt(str.substring(7)) - 1;
+            toDoList.get(ind).unmark();
+            System.out.print("OK, I've marked this task as not done yet:\n" + toDoList.get(ind) + "\n");
+        } else if (str.startsWith("delete")) {
+            if (Integer.parseInt(str.substring(7)) <= 0 || Integer.parseInt(str.substring(7)) > toDoList.size()) {
+                throw new BigdogException("That's out of your list!");
+            }
+            Task temp = toDoList.get(Integer.parseInt(str.substring(7)) - 1);
+            toDoList.remove(Integer.parseInt(str.substring(7)) - 1);
+            System.out.println("Noted. I've removed this task:");
+            System.out.println(temp);
+            System.out.printf("Now you have %s tasks in the list.\n", toDoList.size());
```
Would it be better if you split the code into blocks to improve readability?

Nicholaskoh1 commented on src/main/java/Bigdog.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    public static void main(String[] args) {
+
+        // Scanner object to read user input
+        Scanner scanner = new Scanner(System.in);
+
+        // Welcome message
+        System.out.println("Hello! I'm Bigdog!");
+        System.out.println("What can I do for you?\n");
+
+        fileSaver.loadFromFile(toDoList);
+
+        while (true) {
+
+            // Take in user input
+            String userInput = scanner.nextLine();
+
+            // Terminating condition
+            if (userInput.equalsIgnoreCase("bye")) {
+                System.out.print("Bye. Hope to see you again soon!");
+                break;
+            } else if (userInput.equalsIgnoreCase("list")) {
+
+                // Print the list
+                System.out.println("Here are the tasks in your list:");
+                for (int i = 0; i < toDoList.size(); i++) {
+                    System.out.println((i + 1) + "." + toDoList.get(i));
+                }
+            } else {
+                try {
+                    editList(userInput);
+                } catch (BigdogException e) {
+                    System.out.println(e.getMessage());
+                }
+            }
+        }
+        scanner.close();
```
I like how you separated the code into blocks and added comments!

Nicholaskoh1 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,88 @@
+        this.end = end;
+    }
+
+    public static Deadline of(String s) throws BigdogException {
```
Perhaps you could add JavaDocs here? Along with the other public methods.

Nicholaskoh1 commented on src/main/java/FileSaver.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.ArrayList;
+import java.io.FileNotFoundException;
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.FileReader;
+import java.io.BufferedReader;
```
I like how you didn't just use * to import everything.

Nicholaskoh1 reviewed (commented):
Just some nit picks on the coding standard.

weibinlim committed:
Add Parser.java

Add Storage.java

Add TaskList.java

Add Ui.java

Bigdog.java: Not following OOP

To separate the different components of the chatbot

Add 4 new files, TaskList.java, Ui.java, Storage.java, Parser.java

To separate the 4 components interactions with the user, loading and saving tasks from file, making sense of IO command, operations regarding task list

Delete FileSaver.java

Same functionality as Storage.java

Todo.java: Edit index of substrings

To accommodate for the change of input format

Deadline.java: Edit index of substrings

To accommodate for the change of input format

Event.java: Edit index of substrings

To accommodate for the change of input format

Task.java: Change Task class to be abstract

To ensure that it cannot be initialised to follow OOP

weibinlim committed:
Bigdog.txt: Clear the storage file

weibinlim committed:
Merge branch 'A-MoreOOP'

- - no-ff

weibinlim committed:
Move Biddog.java to package bigdog

Move BigdogException.java to package bigdog

Move Deadline.java to package bigdog

Move Event.java to package bigdog

Move Parser.java to package bigdog

Move Storage.java to package bigdog

Move Task.java to package bigdog

Move TaskList.java to package bigdog

Move Todo.java to package bigdog

Move Ui.java to package bidog

weibinlim committed:
Bigdog.txt: Clear the storage file

weibinlim committed:
Merge branch 'A-Packages'

- - no-ff

weibinlim committed:
Merge branch 'A-Gradle'
 - - no-ff

weibinlim committed:
BigdogException.java: Add to package bigdog by adding the line "package bigdog;"

Deadline.java: Add to package bigdog by adding the line "package bigdog;"

Event.java: Add to package bigdog by adding the line "package bigdog;"

Parser.java: Add to package bigdog by adding the line "package bigdog;"

Storage.java: Add to package bigdog by adding the line "package bigdog;"

Task.java: Add to package bigdog by adding the line "package bigdog;"

TaskList.java: Add to package bigdog by adding the line "package bigdog;"

Todo.java: Add to package bigdog by adding the line "package bigdog;"

Ui.java: Add to package bigdog by adding the line "package bigdog;"

Bigdog.java: Add to package bigdog by adding the line "package bigdog;"

It was a mistake that I forgot to add in the A-Package branch

weibinlim committed:
Setup Gradle to build and run Bigdog project

weibinlim committed:
Bigdog.txt: Add 2 events to storage file

Deadline.java: Edit output for BigdogException throwing

Event.java: Delete line used for testing

TaskList.java: Edit output for BigdogException throwing

TaskList.java: Edit output for method mark(...)

weibinlim committed:
Add BigdogTest.java for JUnit test

Add StorageTest.java for JUnit test

Add StorageTest.txt for JUnit test

Add TastListTest.java for JUnit test

weibinlim committed:
BigdogException: Edit call to superclass constructor

To change error output to move to nextline

weibinlim committed:
Merge branch 'A-JUnit'

- - no-ff

weibinlim committed:
Edit build.gradle file to add parameters for creation of .jar file

weibinlim committed:
Add JavaDocs for all files

weibinlim committed:
Edit all files to follow SE-EDU Java coding standard

Bigdog.java: Delete unused imports

weibinlim committed:
Bigdog.java: Edit run() method to consider case user command is "find"

Bigdog.java: Delete unused imports

Parser.java: Add "find" to VALID_COMMANDS constant

TaskList.java: Add find(...) method

TaskListTest.java: Added JUnit tests for new find(...) method

weibinlim committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/bigdog/Event.java
#	src/main/java/bigdog/Task.java
#	src/test/java/bigdog/StorageTest.java

- - no-ff

weibinlim committed:
Merge branch 'branch-level-9'

- - no-ff

weibinlim committed:
Parser.java: Merge conflict handling manual

TaskList.java: Merge conflict handling manual

btbrandon commented on src/main/java/bigdog/BigdogException.java:
```diff
@@ -0,0 +1,15 @@
+package bigdog;
+
+public class BigdogException extends RuntimeException {
```
you may wish to create more differentiated exceptions for different inputs?

btbrandon commented on src/main/java/bigdog/Task.java:
```diff
@@ -0,0 +1,65 @@
+                    "8. bye";
+
+    public Task(String taskRep, boolean marked) {
+        this.marked = marked;
```
isMarked?

btbrandon commented on src/main/java/bigdog/Bigdog.java:
```diff
@@ -0,0 +1,99 @@
+
+        ui.greet();
+        boolean toContinue = true;
+
```
you may wish to format the spacings more consistently (double spacing here)

btbrandon commented on src/main/java/bigdog/Bigdog.java:
```diff
@@ -0,0 +1,99 @@
+
+        while (toContinue) {
+            try {
+
```
spacing formatting is inconsistent here as well

btbrandon reviewed (commented):
lgtm!

weibinlim committed:
Add checkstyle

weibinlim committed:
Fix coding style violations

weibinlim committed:
Clear the storage file

weibinlim committed:
Merge branch 'branch-A-CheckStyle'

- - no-ff

weibinlim committed:
Add Gui.java

Add Launcher.java

Edit build.gradle to include repositories and dependencies

Edit build.gradle application section

weibinlim committed:
Create GUI for Bigdog

Move GUI package to bigdog package

weibinlim committed:
Implement functional GUI for Bigdog

Echoes user input

weibinlim committed:
Use FXML

Rename Gui.java to Main.java

weibinlim committed:
Fix bug due to name change

weibinlim committed:
Tweak the GUI

Link GUI to the Bigdog chatbot and format it using JavaFX css

weibinlim committed:
Write Javadocs for new classes

Edit Javadocs for existing classes

weibinlim committed:
Task creation and handling classes: Remove unnecessary BigdogException

To allow more efficient bug-fixing, removed unnecessary throws of
BigdogException

Changed methods to use the assert feature instead of throwing exceptions

Mainly edited factory methods and stringToDate(...)

weibinlim committed:
Improve Code Quality

Current code has a few code quality issues such as magic numbers, incomplete JavaDocs and insufficient SLAP.

Add constant variables for Deadline, Event and Storage classes

Improve JavaDocs for Todo, TaskList and Bigdog classes

Improve SLAP for Event, Deadline and Parser classes

Delete unused show() method in TaskList class

Delete unused run() method in Bigdog class

weibinlim committed:
Merge pull request #2 from weibinlim/branch-A-Assertions

Use Java Assertions for error handling

weibinlim committed:
Merge branch 'master' into branch-A-CodeQuality

weibinlim committed:
Merge pull request #3 from weibinlim/branch-A-CodeQuality

Improve Code Quality

weibinlim committed:
Add "view" command for daily schedule, implement task date checks

- Added support for the "view" command to display tasks due on a
  specific day in `Bigdog.java`.
- Implemented `isOnDay` method in `Deadline.java`, `Event.java`, and
  `Todo.java` to check if a task is due on the given date.
- Updated `TaskList.java` to handle schedule viewing and return a
  message if no tasks are found for the given day.
- Modified `Parser.java` to recognise the new "view" command.

- - no-ff

weibinlim committed:
Add JavaDocs for viewSchedule method in Todo, Deadline and Event classes

Bug fix for parse method in Parser class

weibinlim committed:
Merge pull request #4 from weibinlim/branch-B-ViewSchedule

Branch b view schedule

weibinlim committed:
Add Bigdog.txt for external file storage

Enable assertions on build.gradle

Add more error handling capabilities in Bigdog.java

weibinlim committed:
Refactor Deadline task parsing logic and improve error handling

- Replaced manual substring extraction with split() for cleaner parsing
- Improved data corruption checks by validating the number of parts
- Removed unnecessary exception throw after parsing logic update

Fix isOnDay method to compare only date part for task deadlines

- Updated isOnDay() to compare LocalDate values, ignoring time
- Ensures task is considered only if it falls on the exact day

Fix Event isOnDay logic to compare only dates and improve range check

- Updated isOnDay() to use LocalDate for date-only comparisons
- Refined start and end date checks for events to ensure proper range evaluation

weibinlim committed:
Refactor Storage class to use File object instead of String for file path

- Replaced String filePath with File object for better file handling
- Updated constructor to initialise filePath as a File object

Add detailed error message for unrecognised commands in Parser

- Introduced VALID_COMMANDS array for valid task commands.
- Added UNRECOGNIsED_KEYWORDS_MESSAGE to guide users with valid command options.
- Enhanced exception handling to include detailed error message when an invalid argument is encountered.

Refine assertion in Todo.of() to prevent empty todo creation

- Modified the assertion to ensure the input string is not "NA".

Refine method viewSchedule in TaskList class

- Made counter increment after every task on the day identified

weibinlim committed:
Add user guide and UI screenshot

weibinlim committed:
Add Ui screenshot to User Guide

weibinlim committed:
Refactor file path for Bigdog storage

- Removed dynamic file path resolution for Bigdog.txt.
- Set a fixed path to "src/main/java/bigdog/Bigdog.txt" for consistent file access.

weibinlim committed:
Add shutdown functionality to close the application on 'bye' command

- Implemented `Platform.exit()` to close the JavaFX application when user input is "bye".
- Ensured the application exits cleanly after handling the 'bye' command.

weibinlim committed:
Remove reference to Duke in documentation

weibinlim committed:
Rename package bigdog.GUI to bigdog.gui to fit SE-EDU coding standard

weibinlim committed:
Edit main class of the application to the changed package name bigdog.gui

weibinlim committed:
Refactor user guide to be more concise and update Ui image

weibinlim committed:
Bigdog.java: delete unused main() method

Bigdog.txt added new task for Ui image

weibinlim committed:
Refactor code to fit SE-EDU coding standard

- Rename boolean withTime to isTimeIncluded in Deadline and Event classes
- Refactor package name bigdog.GUI to bigdog.gui

weibinlim committed:
Add Ai.md for documentation of AI usage

weibinlim committed:
Merge branch 'branch-A-AiAssisted'

- - no-ff

weibinlim committed:
build: Edit code to allow jar file to work in separate empty folder

- Add main method for Bigdog.java
- Delete additional fields in build.gradle
- Refactor fxml files to run on Java 17 version rather than 22

weibinlim committed:
fix: Unable to save into local hard disk when jar file was in an empty folder

- Edit load() method in storage class to create new file if not found
- Edit instantiation of Bigdog object with new filepath

weibinlim committed:
Organise external files used to save TaskList data

weibinlim committed:
Include more AI documentation

weibinlim committed:
Update User Guide

Run tests on external storage files

weibinlim committed:
Refactor to fit coding standards

weibinlim committed:
Improve error handling for Deadline, Event and TaskList classes

weibinlim committed:
Delete irrelevant test
